### ISSUE TITLE
Questions about Fix up Batch Normalization Layer.

### ISSUE THREAD

**Issue by CGLemon** [lightvector/KataGo#651]
KataGo remove all Batch Normalization layers and combine with gradient clipping. I try to implement it on my project. But I failed many times. Here is my network description.

1. Data Set: KataGO self-play games. Including board size from 7x7 to 19x19. The major rule is Chinese-like only.
2. Network is SE-network. It can predict policy, opp's policy, ownermap, win-draw-loss, side-to-move winrate and score lead.
3. Training on multiple board sizes via masking.
4. Gradient clipping.
5. Re-initialize first conv and last conv of each resnet block if we use the fixup.

Here is the fixup case after training 70k steps(size is 6 block, 128 channels). It looks not bad.

![Screenshot from 2022-06-10 01-15-24](https://user-images.githubusercontent.com/39462842/172905991-27ba22b6-c297-41c9-9601-c778c58521e7.png)

But the performance is very bad if the we input the uncommon board position. Here is the example

![Screenshot from 2022-06-10 01-11-22](https://user-images.githubusercontent.com/39462842/172905322-1522cbfd-f476-447a-94bd-28e4a25418f9.png)

 Here is the using Batch Normalization case after training 784k steps(size is 15 block, 192 channels). We can compare it with fixup case. 

![Screenshot from 2022-06-10 01-20-19](https://user-images.githubusercontent.com/39462842/172906865-b93588c9-5a4e-4587-8278-7801be8fddbe.png)

I have no idea about it. Need some help.

**Comment by lightvector** [comment]
Batch normalization might help to regularize the learning a bit in a way that fixup doesn't, it's certainly possible.

In your example above the networks differ not just in fixup vs batch norm, but they also differ tremendously in network size (6 vs 15 blocks) and in amount of training (70k vs 784k). Do you have any comparison where these parameters are kept equal?

**Comment by CGLemon** [comment]
I already trained some large fixup networks (15 blocks, 800k steps) 2 month ago. It have a same problem with current weights. The difference is that I add some tricks(gradient clipping. initialize first conv and last conv) for current weighs. Seem the trick is not helpful.

I sorry about that. I don't have enough GPU to train same blocks network many times.

**Comment by lightvector** [comment]
Based on recent experiments, I think adding batch norm does help and that it was possibly a mistake to rely entirely on fixup and if you want to continue using batch norm only, you're doing nothing wrong. However, batch norm still has some infrastructurally awkward properties, so some of the latest pytorch experimental code tries a compromise that I'll try to write up later.

Your results are still worse than I'd expect. When you're using fixup, you're following all the recommendations?
* Initializing the first conv weights of each block to 1/sqrt(#blocks) times what the normal xavier or he initialization would want
* Initializing the second conv weights to 0
* In place of each normalization you still have a trainable bias term (i.e. the "beta" of the normalization), or that your convolutions have a trainable bias term.
* You have an extra trainable scalar multiplier ("gamma") somewhere in each block.
* You tune the learning rate for fixup separately from the learning rate you tune for batch norm (they might need different learning rates!)

I've not tried fixup with SE blocks, so I don't know if there's a bad interaction there, that's also possible.

Also, if you just want to make a controlled comparison, but you don't have much GPU power, then you don't necessarily need to train fixup on 15 blocks to 800k steps again, you could also train batch norm on 6 blocks to 70k steps. As long as the comparison is exactly matching on both sides except for one difference, you can learn something about that one difference.

**Comment by CGLemon** [comment]
Every things looks fine, thank you for your help! I close this issue.
